pageserver/src/http/routes.rs

```diff
@@ -1685,6 +1686,45 @@ async fn update_tenant_config_handler(
     json_response(StatusCode::OK, ())
 }
 
+async fn patch_tenant_config_handler(
+    mut request: Request<Body>,
+    _cancel: CancellationToken,
+) -> Result<Response<Body>, ApiError> {
+    let request_data: TenantConfigPatchRequest = json_request(&mut request).await?;
+    let tenant_id = request_data.tenant_id;
+    check_permission(&request, Some(tenant_id))?;
+
+    let state = get_state(&request);
+
+    let tenant_shard_id = TenantShardId::unsharded(tenant_id);
+
+    let tenant = state
+        .tenant_manager
+        .get_attached_tenant_shard(tenant_shard_id)?;
+    tenant.wait_to_become_active(ACTIVE_TENANT_TIMEOUT).await?;
+
+    let base = tenant.get_tenant_conf();
+    let new_tenant_conf = base
+        .apply_patch(request_data.config)
+        .map_err(ApiError::BadRequest)?;
+
+    // This is a legacy API that only operates on attached tenants: the preferred
+    // API to use is the location_config/ endpoint, which lets the caller provide
+    // the full LocationConf.
+    let location_conf = LocationConf::attached_single(
+        new_tenant_conf.clone(),
+        tenant.get_generation(),
+        &ShardParameters::default(),
+    );
+
+    crate::tenant::Tenant::persist_tenant_config(state.conf, &tenant_shard_id, &location_conf)
+        .await
+        .map_err(|e| ApiError::InternalServerError(anyhow::anyhow!(e)))?;
+    tenant.set_new_tenant_config(new_tenant_conf);
```

Comment:
This request is only called from the control plane and that only one operation for a tenant at any given time.
But yeah, it's kind of racy. We wouldn't lose an in-mem update due to the RCU pattern in `set_new_tenant_config`, but we can end up with an on disk config that doesn't match the in-memory one.

We could put the tenant config behind a read/write lock.

Reply:
> We wouldn't lose an in-mem update due to the RCU pattern in set_new_tenant_config

We would, because the patch is applied to a config from a prior `get_tenant_conf()`, not the one that's read inside the RCU. We need to apply the patch inside the RCU closure.

> we can end up with an on disk config that doesn't match the in-memory one

Yeah, I think that's true even for `get_tenant_config_handler()`.

> We could put the tenant config behind a read/write lock.

Do you mean instead of an ArcSwap? Using both would be redundant. I assume we use the ArcSwap because we need the performance?

We could write the file inside the RCU, assuming we use a tempfile and atomic rename, although that's still racy in case of an ill-timed crash. We could take out a Posix lock on the file I suppose. Maybe that's what you meant?

In any case, I'd be fine to defer the file race to a separate PR since it's a pre-existing issue, but I'd be inclined to fix the in-memory race by patching inside an RCU.

Reply:
> We would, because the patch is applied to a config from a prior get_tenant_conf(), not the one that's read inside the RCU. We need to apply the patch inside the RCU closure.

Indeed - will fix.

> In any case, I'd be fine to defer the file race to a separate PR since it's a pre-existing issue

Let's do that.

Reply:
I moved the patch application inside the rcu in https://github.com/neondatabase/neon/pull/10020/commits/86a1f95768f74f205c40b96bde5b421f48dff2c1.

This is still not entirely safe from memory races since we can race on whatever happens after the RCU itself.

---

I played around with a couple of ways of making this completely safe:
* async RwLock around the in-mem and on-disk update. This asyncifies a lot of simple getters, not desirable
* Marking the tenant `InProgress` via `TenantMgr`. This would work, but it's not orthogonal to the current design. We actually rely on the "no-concurrent management requests" property to optimise location config updates. If that doesn't hold we have bigger problems.

Essentially, the concern runs deeper than it initially seems and if we want to address this we should do it in a generic way.

Reply:
> Did we have to implement our own RCU though? Why couldn't we just use the built-in one?

The `update` closure may return an error and the [`rcu`](https://docs.rs/arc-swap/latest/arc_swap/struct.ArcSwapAny.html#method.rcu) interface doesn't support that.

Reply:
Annoying. There's an issue for try_rcu: https://github.com/vorner/arc-swap/issues/94. We could snatch the extension trait from there.

Reply:
Sure. Done in https://github.com/neondatabase/neon/pull/10020/commits/180e806a534f2cf5a8a11365ddc5d8f47bd2543d